docs: add Fees section and align docs with kernel fee logic (#1728)
0xMiden/miden-base: docs/src/fees.md

# Fees

Miden transactions pay a fee that is computed and charged automatically by the transaction kernel during the epilogue.

## How fees are computed

- The fee depends on the number of VM cycles the transaction executes and grows logarithmically with that count. The kernel estimates verification cycles by taking log2 of the estimated total cycles, rounding up to the next power of two. The fee amount is then computed as:
  - `fee_amount = verification_base_fee * ceil(log2(estimated_cycles_rounded_up_to_pow2))`
- The `verification_base_fee` comes from the current reference block’s fee parameters.

Relevant kernel logic:
```246:270:crates/miden-lib/asm/kernels/transaction/lib/epilogue.masm
#! - fee_amount is the computed fee amount of the transaction in the native asset.

> proc.compute_fee

Reviewer: We usually don't copy code into the docs, mainly because the docs should explain things at a higher-level than code and also because copied code goes stale very fast. So I would remove all copied code here.

# Fees

Miden transactions pay a fee that is computed and charged automatically by the transaction kernel during the epilogue.

## How fees are computed

- The fee depends on the number of VM cycles the transaction executes and grows logarithmically with that count.
- The kernel estimates the number of verification cycles by taking log2 of the estimated total execution cycles (rounded up). The result is then multiplied by the `verification_base_fee` from the reference block’s fee parameters.
- In other words, the fee is proportional to the logarithm of the transaction’s number of execution cycles, scaled by the base verification fee defined in the block header.

## Which asset is used to pay fees

- Fees are paid in the chain’s native asset, defined by the current reference block’s fee parameters.

> - The native asset is chosen once as part of the genesis block and then copied to every newly created block, which means the native asset is the same per network during the epilogue.

Reviewer: ```suggestion
- The native asset is chosen once as part of the genesis block and then copied to every newly created block, which means the native asset stays consistent for a given network.
```

## How fees are paid

> - Practically, users should ensure their account’s vault holds sufficient balance of the native asset of the current reference block to cover the fee. The fee is charged automatically; no explicit fee-sending step is required. 

Reviewer: ```suggestion
- Users should ensure their account’s vault holds sufficient balance of the native asset to cover the fee. The fee is charged automatically; no explicit transaction kernel API must be called. 
```

# Fees

Miden transactions pay a fee that is computed and charged automatically by the transaction kernel during the epilogue.

## How fees are computed

- The fee depends on the number of VM cycles the transaction executes and grows logarithmically with that count.
- The kernel estimates the number of verification cycles by taking log2 of the estimated total execution cycles (rounded up). The result is then multiplied by the `verification_base_fee` from the reference block’s fee parameters.

> - In other words, the fee is proportional to the logarithm of the transaction’s execution cost, scaled by the base verification fee defined in the block header.

Reviewer: ```suggestion
- In other words, the fee is proportional to the logarithm of the transaction’s number of execution cycles, scaled by the base verification fee defined in the block header.
```

## Which asset is used to pay fees

> - Fees are paid in the chain’s native asset, defined by the current reference block’s fee parameters.
> - During the epilogue, the kernel constructs the fee asset using the native asset ID and removes it from the account’s vault.
> - If the account does not contain enough of the native asset to cover the computed fee, the transaction fails during the epilogue.

Reviewer: ```suggestion
- Fees are paid in the chain’s native asset, defined by the current reference block’s fee parameters.
- The native asset is chosen once as part of the genesis block and then copied to every newly created block, which means the native asset is the same per network during the epilogue.
```

I'd add a note for how the native asset is chosen and remove the other implementation details.

## Where the fee appears in outputs

- The transaction kernel outputs the computed fee as a fungible asset, and the transaction outputs include it explicitly.

> - Practically, users should ensure their account’s vault holds sufficient balance of the native asset of the current reference block to cover the fee. The fee is charged automatically; no explicit fee-sending step is required. 

Reviewer: I'd rename the section to "How fees are paid" and remove the part about transaction outputs.

Here we can also add
> If the account does not contain enough of the native asset to cover the computed fee, the transaction fails during the epilogue.
